### PR TITLE
UBLOX cellular UDP and TCP api's

### PR DESCRIPTION
**Pull request by mudassar-ublox**
### Description

This PR updates the cellular API's for the u-blox targets. Some u-blox targets must use an internal IP stack while others can use both internal and external IP stacks.  This PR introduces support for the internal IP stack; a new cellular target type, called UBLOX_AT, is added in "CellularTargets.h" and AT classes are added in targets/UBLOX. The API to the internal IP stack supports both UDP and TCP socket operations.

Supported Targets: UBLOX_C030_U201
Tool Chains: IAR, ARM, GCC_ARM

### Pull request type

    [ ] Fix
    [x] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

### CODE REVIEW

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+{
+    return requested_stack == IPV4_STACK ? true : false;
+}
+
+bool UBLOX_AT_CellularNetwork::has_registration(RegistrationType reg_type)
+{
+    return (reg_type == C_REG || reg_type == C_GREG);
+}
+
+nsapi_error_t UBLOX_AT_CellularNetwork::set_access_technology_impl(RadioAccessTechnology opRat)
```
Comment (RobMeades) [inline features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp:55]: Do we need to be more specific here?  For instance, if this SARA-U201 (C030) or SARA-G350 (C027) then should setting the RAT to CATM1 or NB1 return an error?  Likewise the other way around for R410M and N2xx.

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+    int a,b;
+    CellularSocket *socket;
+
+    a =_at.read_int();
+    b =_at.read_int();
+
+    socket = find_socket(a);
+    if (socket != NULL) {
+        socket->rx_avail = true;
+        pendingBytes = b;
```
Comment (RobMeades) [inline features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp]: Doesn't the `pendingBytes` value need to be stored per socket (otherwise it will be confused if we have more than one socket running at the same time)?

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+    clear_socket(socket);
+}
+
+
+int UBLOX_AT_CellularStack::get_max_socket_count()
+{
+    return UBLOX_MAX_SOCKET;
+}
+
+int UBLOX_AT_CellularStack::get_max_packet_size()
```
Comment (RobMeades) [inline features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp]: If we return `UBLOX_MAX_PACKET_SIZE` here, aren't we telling the user that they can send no more than 1024 bytes per call to `sendto()`/`send()`, when in fact it is possible to send more than that because you "chunk" the data up appropriately?

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+
+nsapi_error_t UBLOX_AT_CellularStack::socket_open(nsapi_socket_t *handle, nsapi_protocol_t proto)
+{
+    if (!is_protocol_supported(proto) || !handle) {
+        return NSAPI_ERROR_UNSUPPORTED;
+    }
+
+    int max_socket_count = get_max_socket_count();
+
+    if (!_socket) {
```
Comment (RobMeades) [inline features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp]: I may have got this wrong, if so my apologies, but looking at the code here this seems to zero all the socket data whereas I would have expected it to open a single socket, add it to the list of open sockets in an array of CellularSocket structs, and leave all the others alone.  How does one open multiple sockets at the same time?

**Review by RobMeades — COMMENTED** [review]
Looking really good Mudassar: my queries now are mainly around how running multiple sockets in parallel works (sorry if I've misunderstood the code in this respect).  I will also add Phil Ware as a reviewer now.

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp (hunk 1 above)
Comment (mudassar-ublox) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp:55]: Then can we do like that 

    switch(opRat) {
#if defined(TARGET_UBLOX_C030_U201) || defined(TARGET_UBLOX_C027)
		case RAT_UTRAN:
			break;
		case RAT_EGPRS:
			break;
		case RAT_E_UTRAN:
			break;
#elif defined(TARGET_UBLOX_C030_R410M)
		case RAT_CATM1:
			break;
#elif defined(TARGET_UBLOX_C030_R410M) || defined(TARGET_UBLOX_C030_N211)
		case RAT_NB1:
			break;
#endif
		default: {
			_op_act = RAT_UNKNOWN;
			return NSAPI_ERROR_UNSUPPORTED;
		}
    }

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp (hunk 2 above)
Comment (mudassar-ublox) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp]: My apologies. It was in my mind, forgot to discuss. Previously SockCtrl structure have pendingByte attribute but now CellularSocket class don't have any attribute for pendingBytes. So how to handle it.

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp (hunk 3 above)
Comment (mudassar-ublox) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp]: Yes we can send more data now. My idea of not removing that function that user can see that whats the max packet size is and if he want to use that limit so that he can use it. What you think we should remove it as we have handled data more than max packet size.

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp (hunk 4 above)
Comment (mudassar-ublox) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp]: Implemented virtual function of AT_CellularStack class and kept its code and added "create_socket_impl" function. This code is checking that _socket are allocated in memory or not, if not then it is creating and initializing all of them with zero.

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp (hunk 1 above)
Comment (RobMeades) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp:55]: Yes, I think something like that would be good.

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp (hunk 2 above)
Comment (RobMeades) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp]: As discussed, we could (as a first step) add a `pendingBytes` field to the `CellularSockets` data structure.  If ARM object to this when we submit the PR then we can ask them for their solution.

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp (hunk 3 above)
Comment (RobMeades) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp]: Let's see how ARM reply to https://github.com/ARMmbed/mbed-os/issues/7181.

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp (hunk 4 above)
Comment (RobMeades) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp]: Ah, yes, that makes perfect sense.

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp (hunk 1 above)
Comment (bqam-ublox) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp:55]: Just thinking so what about C027? 
(Please ignore this comment if not relevant)

Comment (RobMeades) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp:55]: Good point, this does need to work on C027.

Comment (bqam-ublox) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp:55]: I guess the second condition here was meant to be TARGET_UBLOX_C027 (more of a typo) : `#if defined(TARGET_UBLOX_C030_U201) || defined(TARGET_UBLOX_C030_U201)` . 

Point to note, SARA-G350 is 2G only so UTRAN, E-UTRAN should return error so just requires further classification.

Comment (mudassar-ublox) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp:55]: Yes, its typo mistake.

Comment (bqam-ublox) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp:55]: This should also check all values in the enum RadioAccessTechnology : https://github.com/ARMmbed/mbed-os/blob/master/features/cellular/framework/API/CellularNetwork.h#L119. 
Otherwise, for example, RAT_HSDPA/RAT_HSUPA would return error although they are valid for SARA-U201.

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+
+nsapi_size_or_error_t UBLOX_AT_CellularStack::socket_sendto(nsapi_socket_t handle, const SocketAddress &addr, const void *data, unsigned size)
+{
+    CellularSocket *socket = (CellularSocket *)handle;
+    if (!socket) {
+        return NSAPI_ERROR_DEVICE_ERROR;
+    }
+
+    nsapi_size_or_error_t ret_val = NSAPI_ERROR_OK;
+
```
Comment (philwareublox) [inline features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp]: The create socket section below this comment, why is this required? 
I'm not sure if we want to create a socket if it hasn't been - this should have been done previously. 
The developer should know if his socket is created or not, and if not, return an error. 
Did you have a scenario for this? 

(Also I don't see this in the other functions)

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.h [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            const void *data, nsapi_size_t size);
+
+    virtual nsapi_size_or_error_t socket_recvfrom_impl(CellularSocket *socket, SocketAddress *address,
+            void *buffer, nsapi_size_t size);
+
+    virtual nsapi_size_or_error_t socket_sendto(nsapi_socket_t handle, const SocketAddress &address,
+            const void *data, nsapi_size_t size);
+			
+    virtual nsapi_error_t socket_close_impl(int sock_id);
+
```
Comment (philwareublox) [inline features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.h]: Where are the other TCP type Send and Receive functions?
I can only see SendTo and RecvFrom...

**Review by RobMeades — CHANGES_REQUESTED** [review]
Only one comment: for the UDP send case, given Kevin Bracey's answer here: https://github.com/ARMmbed/mbed-os/issues/7181#issuecomment-397596030, I think we should be returning NSAPI_ERROR_PARAMETER if the call to send a UDP packet is for a packet length greater than UBLOX_MAX_PACKET_SIZE.  Otherwise, provided our tests are all passing on C030 and C027 (making the above change will cause some of them to fail, you will need to limit the UDP packet size for them to pass), I think this is good to go.

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.h (hunk 6 above)
Comment (philwareublox) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.h]: I understand how this is implemented now :)

File: features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp (hunk 5 above)
Comment (philwareublox) [reply features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp]: Any thoughts on this comment about creating a socket if there is none? 
If it was TCP "send" function then you would also need to connect() to the server before this "send" was executed. 
Probably best not bother creating a socket if not created by the user, simply return an error.

**Review by RobMeades — APPROVED** [review]
Looks good to me.

**Review by philwareublox — APPROVED** [review]
